File: 3014-minimum-number-of-pushes-to-type-word-i/3014-minimum-number-of-pushes-to-type-word-i.ts
function minimumPushes(word: string): number {
    const freq = new Array(26).fill(0);
    for (const ch of word) {
        freq[ch.charCodeAt(0) - 97]++;
    }
    freq.sort((a, b) => b - a); // descending

    let res = 0;
    for (let i = 0; i < freq.length; i++) {
        if (freq[i] === 0) break;

        if (i < 8) res += freq[i];

        else if (i < 16) res += freq[i] * 2;

        else if (i < 24) res += freq[i] * 3;
        
        else res += freq[i] * 4;
    }
    return res;
}

File: 3014-minimum-number-of-pushes-to-type-word-i/3014-minimum-number-of-pushes-to-type-word-i.rs
impl Solution {
    pub fn minimum_pushes(word: String) -> i32 {
        let mut freq = [0; 26];
        for b in word.bytes() {
            freq[(b - b'a') as usize] += 1;
        }
        freq.sort_unstable_by(|a, b| b.cmp(a)); // descending

        let mut res = 0;
        for (i, &cnt) in freq.iter().enumerate() {
            if cnt == 0 { break; }
            res += if i < 8 {
                cnt
            } else if i < 16 {
                cnt * 2
            } else if i < 24 {
                cnt * 3
            } else {
                cnt * 4
            };
        }
        res
    }
}